homeassistant/components/openrgb/coordinator.py

```diff
@@ -0,0 +1,141 @@
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=15),
+            config_entry=config_entry,
+            request_refresh_debouncer=Debouncer(
+                hass, _LOGGER, cooldown=0.5, immediate=False
+            ),
+        )
+        self.host = config_entry.data[CONF_HOST]
+        self.port = config_entry.data[CONF_PORT]
+        self.entry_id = config_entry.entry_id
+        self.server_address = f"{self.host}:{self.port}"
+        self.client_lock = asyncio.Lock()
+
+        config_entry.async_on_unload(self.async_client_disconnect)
+
+    async def _async_setup(self) -> None:
+        """Set up the coordinator by connecting to the OpenRGB SDK server."""
+        self.client = await self.hass.async_add_executor_job(
+            OpenRGBClient,
+            self.host,
+            self.port,
+            DEFAULT_CLIENT_NAME,
+        )
+
+    async def _async_update_data(self) -> dict[str, Device]:
+        """Fetch data from OpenRGB."""
+        async with self.client_lock:
+            try:
+                await self.hass.async_add_executor_job(self._client_update)
+            except CONNECTION_ERRORS as err:
+                raise UpdateFailed(
+                    translation_domain=DOMAIN,
+                    translation_key="communication_error",
+                    translation_placeholders={
+                        "server_address": self.server_address,
+                        "error": str(err),
+                    },
+                ) from err
+
+        # Return devices indexed by their key
+        return {self._get_device_key(device): device for device in self.client.devices}
+
+    def _client_update(self) -> None:
+        try:
+            self.client.update()
+        except CONNECTION_ERRORS:
+            # Try to reconnect once
+            self.client.disconnect()
+            self.client.connect()
+            self.client.update()
+
+    def _get_device_key(self, device: Device) -> str:
+        """Build a stable device key.
+
+        Note: the OpenRGB device.id is intentionally not used because it is just
+        a positional index that can change when devices are added or removed.
+        """
+        parts = (
+            self.entry_id,
+            device.type.name,
+            device.metadata.vendor or "none",
+            device.metadata.description or "none",
+            device.metadata.serial or "none",
+            device.metadata.location or "none",
+        )
+        # Double pipe is readable and is unlikely to appear in metadata
+        return "||".join(parts)
+
+    async def async_client_disconnect(self, *args) -> None:
+        """Disconnect the OpenRGB client."""
+        if not hasattr(self, "client"):
+            # If async_config_entry_first_refresh failed, client will not exist
+            return
```

Comment:
but if the first refresh failed, I don't believe we call an unload right?

Reply:
So, turns out `unload` is called even when the first refresh fails.

If I turn off my OpenRGB SDK Server and start Home Assistant with a breakpoint in this line, it will trigger right away.